code/CMSEditLinkExtension.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace SilverStripe\Admin;
+
+use LogicException;
+use SilverStripe\CMS\Controllers\CMSMain;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\Director;
+use SilverStripe\Core\Extension;
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\GridField\GridField;
+use SilverStripe\Forms\GridField\GridFieldDetailForm;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * An extension that automatically generates a CMS edit link for DataObjects even if
+ * they are canonically edited in some nested {@link GridField}.
+ * Designed to be used in conjunction with the {@link CMSPreviewable} interface.
+ *
+ * For nested relations (e.g. a DataObject managed in a GridField of another DataObject)
+ * you can apply this extension to both the parent and the child object and the links
+ * will chain down the nested `GridField`s to the root cms edit owner.
+ *
+ * You must set a cms_edit_owner config variable which defines the cms edit
+ * owner for this class.
+ * e.g. set this to a {@link LeftAndMain} class:
+ * private static string cms_edit_owner = MyModelAdmin::class;
+ * or to a has_one relation:
+ * private static string cms_edit_owner = 'Parent';
+ *
+ * Note that the cms edit owner must implement a getCMSEditLinkForManagedDataObject() method.
+ *
+ * If the cms edit owner is a has_one relation, the class on the other end
+ * of the relation must have a CMSEditLink() method.
+ */
+class CMSEditLinkExtension extends Extension
+{
+    private static string $cms_edit_owner = '';
+
+    /**
+     * Get the admin or DataObject which owns this object for CMS editing purposes.
```

Comment:
```suggestion
     * Get the ModelAdmin, LeftAndMain or DataObject which owns this object for CMS editing purposes.
```

Reply:
ModelAdmin _is_ a LeftAndMain - and they're all admin. I'll make this change but it's super subjective.

Reply:
done